qualtran/cirq_interop/_cirq_interop.py

```diff
@@ -188,67 +204,97 @@ def cirq_optree_to_cbloq(
 
+    cirq_quregs = {k: np.apply_along_axis(QReg, -1, v) for k, v in cirq_quregs.items()}
+
     assert signature is not None and cirq_quregs is not None
 
     bb, initial_soqs = BloqBuilder.from_signature(signature, add_registers_allowed=False)
 
-    # Magic to make sure signature of the CompositeBloq matches `Signature`.
-    qubit_to_qvar = {}
+    # 0. Helper functions
+    def _update_qreg_to_qvar(in_reg: QReg, qreg_to_qvar: Dict[QReg, Soquet]) -> None:
+        """Takes care of splits and joins to make sure `qreg_to_qvar[in_reg]` exists."""
+        if in_reg in qreg_to_qvar:
+            # This is the easy case when no split / joins are needed.
+            return
+
+        # Split everything and join s.t. joined register corresponds to `in_reg`.
+        # 1. Split all registers containing at-least one qubit corresponding to `in_reg`.
+        in_reg_qubits = set(in_reg.qubits)
+
+        new_qreg_to_qvar: Dict[QReg, Soquet] = {}
+        for qreg, soq in qreg_to_qvar.items():
+            if len(qreg.qubits) > 1 and any(q in qreg.qubits for q in in_reg_qubits):
+                new_qreg_to_qvar |= {QReg(q): s for q, s in zip(qreg.qubits, bb.split(soq=soq))}
+            else:
+                new_qreg_to_qvar[qreg] = soq
+        qreg_to_qvar.clear()
+        soqs_to_join = []
+        for qreg, soq in new_qreg_to_qvar.items():
+            if len(in_reg_qubits) > 1 and qreg.qubits and qreg.qubits[0] in in_reg_qubits:
+                assert len(qreg.qubits) == 1, "Individual qubits should have been split by now."
+                soqs_to_join.append(soq)
+            else:
+                qreg_to_qvar[qreg] = soq
+        if soqs_to_join:
+            qreg_to_qvar[in_reg] = bb.join(np.array(soqs_to_join))
+
+    def _find_input_soqs(
+        op_quregs: Dict[str, NDArray[QReg]], qreg_to_qvar: Dict[QReg, Soquet]
+    ) -> Dict[str, NDArray[Soquet]]:
+        qvars_in: Dict[str, NDArray[Soquet]] = {}
+        for reg_name, quregs in op_quregs.items():
+            flat_soqs: List[Soquet] = []
+            for qureg in quregs.flatten():
+                _update_qreg_to_qvar(qureg, qreg_to_qvar)
+                flat_soqs.append(qreg_to_qvar[qureg])
+            qvars_in[reg_name] = np.array(flat_soqs).reshape(quregs.shape)
+        return qvars_in
+
+    # 1. Compute qreg_to_qvar.
+    qreg_to_qvar: Dict[QReg, Soquet] = {}
     for reg in signature.lefts():
         assert reg.name in cirq_quregs
         soqs = initial_soqs[reg.name]
         if isinstance(soqs, Soquet):
-            soqs = np.asarray(soqs)[np.newaxis, ...]
-        if reg.bitsize > 1:
-            soqs = np.array([bb.split(soq) for soq in soqs.flatten()])
-        soqs = soqs.reshape(reg.shape + (reg.bitsize,))
-        assert cirq_quregs[reg.name].shape == soqs.shape
-        qubit_to_qvar |= zip(cirq_quregs[reg.name].flatten(), soqs.flatten())
-
-    allocated_qubits = set()
+            soqs = np.asarray(soqs)
```

Comment:
In this case, `np.asarray(soqs)` works similar to `np.array(soqs)` and returns a `array(soquet, dtype=object)` with shape `()`. `np.asarray([soqs])` would instead return an array with shape `(1,)`, which is not what we want. 

Replaced with `np.array(soqs)` to avoid the confusion.